src/networkaddress.rs

```diff
@@ -107,28 +126,27 @@ impl MulticastJoinable for SocketAddrV4 {
         interface: InterfaceName,
         _token: PrivateToken,
     ) -> std::io::Result<()> {
+        let mode = crate::interface::LinuxNetworkMode::Ipv4;
+
         let request = libc::ip_mreqn {
             imr_multiaddr: libc::in_addr {
                 s_addr: u32::from_ne_bytes(self.ip().octets()),
             },
             imr_address: libc::in_addr {
                 s_addr: u32::from_ne_bytes(Ipv4Addr::UNSPECIFIED.octets()),
             },
-            imr_ifindex: InterfaceDescriptor {
-                interface_name: Some(interface),
-                mode: crate::interface::LinuxNetworkMode::Ipv4,
-            }
-            .get_index()
-            .ok_or(std::io::ErrorKind::InvalidInput)? as _,
+            imr_ifindex: get_interface_index(interface, mode)? as _,
         };
-        // Safety:
+
+        // # Safety
+        //
         // value points to a struct of length option_len, of type ip_mreq as expected for IPPROTO_IP/IP_ADD_MEMBERSHIP
         cerr(unsafe {
             libc::setsockopt(
                 socket,
                 libc::IPPROTO_IP,
                 libc::IP_ADD_MEMBERSHIP,
-                &request as *const _ as *const _,
+                &request as *const libc::ip_mreqn as *const libc::c_void,
```

Comment:
explicit is better than implicit. Otherwise you're just guessing types, and copy/pasting might invalidate the assumptions made by the `_`

Reply:
No, the explicit here is just distraction and repeating type names don't magically make them more correct. What the code needs to do is to provide a reference to the request, just encoded as a slightly inconvenient type which the first line gets us to. The reference to request part is what I think the core of this line is, not the unfortunate conversion that is needed.